hive-metastore/src/main/java/org/apache/iceberg/hive/legacy/LegacyHiveSchemaUtils.java

```diff
@@ -0,0 +1,691 @@
+              sourceFieldNameLowercase, isNewlyAddedField);
+          break;
+        case ARRAY:
+          handleArraySchemaMerge(metaDataFieldsMap, fieldAssembler, sourceField, sourceFieldName, sourceFieldSchema,
+              sourceFieldNameLowercase, isNewlyAddedField);
+          break;
+        case RECORD:
+          handleRecordSchemaMerge(metaDataFieldsMap, fieldAssembler, sourceFieldName, sourceFieldSchema,
+              sourceFieldNameLowercase, isNewlyAddedField);
+          break;
+        case UNION:
+          throw new IllegalArgumentException(
+              "We do not support UNION type except nullable field which is handled in other types");
+        default:
+          throw new IllegalArgumentException("Unsupported Schema type: " + sourceFieldSchema.getType().toString());
+      }
+    }
+
+    return fieldAssembler.endRecord();
+  }
+
+  private static void generateMetaDataFieldsMap(List<Schema.Field> metaDataFields,
+      Map<String, Queue<Schema.Field>> metaDataFieldsMap) {
+    for (Schema.Field field : metaDataFields) {
+      // The field name of sourceSchema is from hive thus case sensitivity info can be lost in some cases
+      // We use lowercase field name to check whether a field is an metaData field or source field
+      String fieldLowercase = field.name().toLowerCase();
+      if (metaDataFieldsMap.containsKey(fieldLowercase)) {
+        metaDataFieldsMap.get(fieldLowercase).offer(field);
+      } else {
+        Queue<Schema.Field> fieldsQueue = new LinkedList<>();
+        fieldsQueue.offer(field);
+        metaDataFieldsMap.put(fieldLowercase, fieldsQueue);
+      }
+    }
+  }
+
+  private static void handleMapSchemaMerge(Map<String, Queue<Schema.Field>> metaDataFieldsMap,
+      SchemaBuilder.FieldAssembler<Schema> fieldAssembler, Schema.Field sourceField, String sourceFieldName,
+      Schema sourceFieldSchema, String sourceFieldNameLowercase, boolean isNewlyAddedField) {
+    if (isNewlyAddedField) {
+      fieldAssembler.name(sourceFieldName).type().optional().map().values(sourceFieldSchema.getValueType());
+    } else {
+      Schema.Field oldField = metaDataFieldsMap.get(sourceFieldNameLowercase).poll();
+      Schema oldFieldSchema = oldField.schema();
+
+      Schema evolvedMapSchema = mergeArrayAndMapFieldSchema(oldFieldSchema, sourceFieldSchema, Schema.Type.MAP);
+      Schema.Field evolvedMapField =
+          new Schema.Field(oldField.name(), evolvedMapSchema, oldField.doc(), oldField.defaultVal());
+
+      mergeAndAppendField(sourceField, evolvedMapField, fieldAssembler);
+    }
+  }
+
+  private static void handleArraySchemaMerge(Map<String, Queue<Schema.Field>> metaDataFieldsMap,
+      SchemaBuilder.FieldAssembler<Schema> fieldAssembler, Schema.Field sourceField, String sourceFieldName,
+      Schema sourceFieldSchema, String sourceFieldNameLowercase, boolean isNewlyAddedField) {
```

Comment:
I am not able to follow what is going on in this method, especially with the new names we keep introducing: evolved, old, etc. We have two schemas: hive and supplemental. Let us reuse the names whenever possible.